### PR TITLE
Remove `fprintf` from the library

### PR DESCRIPTION
**Pull request by folkertdev**
it's still present in the binaries though

### CODE REVIEW

File: blocksort.rs [hunk 1]
```diff
@@ -347,11 +272,7 @@ unsafe extern "C" fn fallbackSort(
     H = 1 as libc::c_int;
     loop {
         if verb >= 4 as libc::c_int {
-            fprintf(
-                stderr,
-                b"        depth %6d has \0" as *const u8 as *const libc::c_char,
-                H,
-            );
+            eprint!("        depth {:>6} has ", H,);
```
Comment (bjorn3) [inline blocksort.rs:275]:
```suggestion
            eprint!("        depth {:>6} has ", H);
```
nit

File: blocksort.rs [hunk 2]
```diff
@@ -1224,10 +1132,8 @@ unsafe extern "C" fn mainSort(
                         as Int32;
                     if hi > lo {
                         if verb >= 4 as libc::c_int {
-                            fprintf(
-                                stderr,
-                                b"        qsort [0x%x, 0x%x]   done %d   this %d\n\0" as *const u8
-                                    as *const libc::c_char,
+                            eprintln!(
+                                "        qsort [0x{:x}, 0x{:x}]   done {}   this {}",
```
Comment (bjorn3) [inline blocksort.rs:1136]:
```suggestion
                                "        qsort [{:#x}, {:#x}]   done {}   this {}",
```
geeft dezelfde uitvoer.

File: bzip2.rs [hunk 3]
```diff
@@ -7,6 +7,10 @@
 #![allow(unused_mut)]
 
 use bzip2 as _;
```
Comment (bjorn3) [inline bzip2.rs:9]:
```suggestion
```

File: bzlib.rs [hunk 4]
```diff
@@ -1,39 +1,32 @@
...
-    fn strcmp(_: *const libc::c_char, _: *const libc::c_char) -> libc::c_int;
-    static mut BZ2_rNums: [Int32; 512];
-    static mut BZ2_crc32Table: [UInt32; 256];
-    fn BZ2_compressBlock(_: *mut EState, _: Bool);
-    fn BZ2_decompress(_: *mut DState) -> Int32;
 }
+
+macro_rules! version {
+    () => {
+        "1.1.0"
```
Comment (bjorn3) [inline bzlib.rs:21]:
```suggestion
        env!("CARGO_PKG_VERSION")
```

if you want to use the version from Cargo.toml. Or maybe

```suggestion
        concat!("1.1.0-rs", env!("CARGO_PKG_VERSION"))
```

to indicate that it is based on bzip 1.1.0 while still independently providing a version for the rust port?

File: bzlib.rs [hunk 5]
```diff
@@ -1549,15 +1538,14 @@ pub unsafe extern "C" fn BZ2_bzDecompress(mut strm: *mut bz_stream) -> libc::c_i
             {
                 (*s).calculatedBlockCRC = !(*s).calculatedBlockCRC;
                 if (*s).verbosity >= 3 as libc::c_int {
-                    fprintf(
-                        stderr,
-                        b" {0x%08x, 0x%08x}\0" as *const u8 as *const libc::c_char,
+                    eprint!(
+                        " {{0x{:08x}, 0x{:08x}}}",
```
Comment (bjorn3) [inline bzlib.rs:1542]:
```suggestion
                        " {{{:#08x}, {:#08x}}}",
```

File: bzlib.rs [hunk 6]
```diff
@@ -1574,10 +1562,8 @@ pub unsafe extern "C" fn BZ2_bzDecompress(mut strm: *mut bz_stream) -> libc::c_i
             let mut r: Int32 = BZ2_decompress(s);
             if r == 4 as libc::c_int {
                 if (*s).verbosity >= 3 as libc::c_int {
-                    fprintf(
-                        stderr,
-                        b"\n    combined CRCs: stored = 0x%08x, computed = 0x%08x\0" as *const u8
-                            as *const libc::c_char,
+                    eprint!(
+                        "\n    combined CRCs: stored = 0x{:08x}, computed = 0x{:08x}",
```
Comment (bjorn3) [inline bzlib.rs:1566]:
```suggestion
                        "\n    combined CRCs: stored = {:#08x}, computed = {:#08x}",
```

File: compress.rs [hunk 7]
```diff
@@ -1273,11 +1185,7 @@ unsafe extern "C" fn sendMTFValues(mut s: *mut EState) {
         i += 1;
     }
     if (*s).verbosity >= 3 as libc::c_int {
-        fprintf(
-            stderr,
-            b"selectors %d, \0" as *const u8 as *const libc::c_char,
-            (*s).numZ - nBytes,
-        );
+        eprint!("selectors {}, ", (*s).numZ - nBytes,);
```
Comment (bjorn3) [inline compress.rs:1188]:
```suggestion
        eprint!("selectors {}, ", (*s).numZ - nBytes);
```

File: compress.rs [hunk 8]
```diff
@@ -1300,11 +1208,7 @@ unsafe extern "C" fn sendMTFValues(mut s: *mut EState) {
         t += 1;
     }
     if (*s).verbosity >= 3 as libc::c_int {
-        fprintf(
-            stderr,
-            b"code lengths %d, \0" as *const u8 as *const libc::c_char,
-            (*s).numZ - nBytes,
-        );
+        eprint!("code lengths {}, ", (*s).numZ - nBytes,);
```
Comment (bjorn3) [inline compress.rs:1211]:
```suggestion
        eprint!("code lengths {}, ", (*s).numZ - nBytes);
```

File: compress.rs [hunk 9]
```diff
@@ -1654,11 +1558,7 @@ unsafe extern "C" fn sendMTFValues(mut s: *mut EState) {
         BZ2_bz__AssertH__fail(3007 as libc::c_int);
     }
     if (*s).verbosity >= 3 as libc::c_int {
-        fprintf(
-            stderr,
-            b"codes %d\n\0" as *const u8 as *const libc::c_char,
-            (*s).numZ - nBytes,
-        );
+        eprintln!("codes {}", (*s).numZ - nBytes,);
```
Comment (bjorn3) [inline compress.rs:1561]:
```suggestion
        eprintln!("codes {}", (*s).numZ - nBytes);
```

File: compress.rs [hunk 10]
```diff
@@ -1714,11 +1612,7 @@ pub unsafe extern "C" fn BZ2_compressBlock(mut s: *mut EState, mut is_last_block
         bsPutUChar(s, 0x90 as libc::c_int as UChar);
         bsPutUInt32(s, (*s).combinedCRC);
         if (*s).verbosity >= 2 as libc::c_int {
-            fprintf(
-                stderr,
-                b"    final combined CRC = 0x%08x\n   \0" as *const u8 as *const libc::c_char,
-                (*s).combinedCRC,
-            );
+            eprint!("    final combined CRC = 0x{:08x}\n   ", (*s).combinedCRC,);
```
Comment (bjorn3) [inline compress.rs:1615]:
```suggestion
            eprint!("    final combined CRC = 0x{:08x}\n   ", (*s).combinedCRC);
```

File: decompress.rs [hunk 11]
```diff
@@ -1106,11 +1000,7 @@ pub unsafe extern "C" fn BZ2_decompress(mut s: *mut DState) -> Int32 {
                         (*s).currBlockNo += 1;
                         (*s).currBlockNo;
                         if (*s).verbosity >= 2 as libc::c_int {
-                            fprintf(
-                                stderr,
-                                b"\n    [%d: huff+mtf \0" as *const u8 as *const libc::c_char,
-                                (*s).currBlockNo,
-                            );
+                            eprint!("\n    [{}: huff+mtf ", (*s).currBlockNo,);
```
Comment (bjorn3) [inline decompress.rs:1003]:
```suggestion
                            eprint!("\n    [{}: huff+mtf ", (*s).currBlockNo);
```

**Review by bjorn3 — CHANGES_REQUESTED** [review]

File: bzlib.rs (hunk 4 above)
Comment (folkertdev) [reply bzlib.rs:21]: yeah, we do something clever here in zlib-rs. Let's come back to that later

Comment (folkertdev) [reply bzlib.rs:21]: tracking issue https://github.com/trifectatechfoundation/libbzip2-rs/issues/2

**Pull request merged**